src/analysis/asymptotics.lean

```diff
@@ -1058,6 +1058,101 @@ theorem is_o_iff_tendsto {f g : α → 𝕜} {l : filter α}
   is_o f g l ↔ tendsto (λ x, f x / (g x)) l (𝓝 0) :=
 iff.intro is_o.tendsto_0 (is_o_of_tendsto hgf)
 
+/-!
+### Eventually (u / v) * v = u
+
+If `u` and `v` are linked by an `is_O_with` relation, then we
+eventually have `(u / v) * v = u`, even if `v` vanishes.
+-/
+
+section eventually_mul_div_cancel
+
+variables {u v : α → 𝕜}
+
+lemma is_O_with.eventually_mul_div_cancel (h : is_O_with c u v l) :
+  ∀ᶠ x in l, (u x / v x) * v x = u x :=
+begin
+  rw is_O_with at h,
+  rw eventually_iff_exists_mem at *,
+  rcases h with ⟨s, hsl, hs⟩,
+  use [s, hsl],
+  intros y hy,
+  specialize hs y hy,
+  by_cases hvy : v y = 0,
+  { rw hvy at *,
+    rw [norm_zero, mul_zero] at hs,
+    have hs' := le_antisymm hs (norm_nonneg _),
+    rw norm_eq_zero at hs',
+    rw hs',
+    norm_num },
+  { rw div_mul_cancel _ hvy }
+end
+
+lemma is_O.eventually_mul_div_cancel (h : is_O u v l) : ∀ᶠ x in l, (u x / v x) * v x = u x :=
+let ⟨c, hc⟩ := h in hc.eventually_mul_div_cancel
+
+lemma is_o.eventually_mul_div_cancel (h : is_o u v l) : ∀ᶠ x in l, (u x / v x) * v x = u x :=
+(h zero_lt_one).eventually_mul_div_cancel
+
+end eventually_mul_div_cancel
+
+/-! ### Equivalent definitions of the form `∃ φ, u =ᶠ[l] φ * v` in a `normed_field`. -/
+
+section exists_mul_eq
+
+variables {u v : α → 𝕜}
+
+lemma is_O_with_of_eq_mul (φ : α → 𝕜) (hφ : ∀ᶠ x in l, ∥φ x∥ ≤ c) (h : u =ᶠ[l] φ * v) :
```

Comment:
What's the point of stating this independently from the iff version if it's not even in a namespace? I would rather prove the iff version and then state `is_O_with.eq_mul` for the other direction so that it can be used through the dot notation, ie given `h : is_O_with c u v l ` you could write `rcases h.eq_mul with ⟨φ, hφ, huφ⟩`.

Reply:
This one does not require `0 ≤ c`, which makes it easier to use in the other proofs. Do you want me to mark this lemma as private ?

Also I'll definitely state the other direction using dot notation.

Reply:
Please don't make it private. Instead, copy your explanation to the docstring..